00xx-account-deletion.md

```diff
@@ -0,0 +1,384 @@
+# Account deletion for consumable accounts
+
+## Summary
+
+The capacity to delete accounts is a prerequisite to implement *consumable
+accounts* in order to propose the *zkPromise* abstraction in `o1js`.
+
+## Motivation ##
+
+Account deletion is a key protocol feature with the goal to enable the so-called
+*zkPromise* abstraction in `o1js`.
+
+Account deletion will allow to create as many accounts as you want, **then**
+receive the deposit back when you no longer need them. This is key for zkApps
+developers who want to trigger actions on other accounts in a deadlock-safe way
+so that they can create a cross-account smart contracts & applications.
+
+## Detailed Design
+
+The support for account deletion relies on adding features at the levels of
+transaction logic(s) and storage (ledgers). The latter is a prerequisite of the
+former.
+
+### Storage
+
+Within the protocol-related codebase, the lowest-level support lies in having a
+working and sound `remove` functions for the different ledger implementations
+(`Database`, `Any_ledger`, `Null_ledger`, `Syncable_ledger`) and in masking
+merkle trees.
+
+#### <a name="merkle_trees"></a> Merkle trees
+
+The current implementation of in-memory ledgers use a fixed-depth [Merkle
+tree](https://en.wikipedia.org/wiki/Merkle_tree).
+
+For insertion, the current implementation keeps track of the "fill frontier",
+that is, the leftmost empty slot of the tree. Since removal is not possible, the
+leaves between the leftmost leaf of the Merkle tree and the rightmost-filled
+leaf are *all* filled with data (i.e., non-empty acccounts).
+
+Now, on removal, there are (at least) 2 options:
+
+1. Symbolically mark the location as removed and make this knowledge available
+   at data structure level such that, upon new insertions, these available
+   locations are looked at before inserting on the leftmost available
+   location. This solution extends the datatype of current implementation with a
+   structure tracking freed locations.
+
+2. Switch the value currently indexed on the frontier of the tree with the
+   location that will be removed, a little bit like one does when implementing a
+   heap data structure. This solution keeps the insertion of new data to the
+   leftmost available locations and thus solution wouldn't change the current
+   datatype.
+
+We will only discuss option 1 here and delay the discussion about option 2 in
+[the alternative section](#alternative).
+
+
+#### Example scenario
+
+To illustrate option 1, let us assume we have the following Merkle tree of depth
+2, with an empty free list. This Merkle tree has one location marked with `x` -
+this marks the empty account.
```

Comment:
I don't think this is the design we want in practice: if we just replace it with the empty account, we don't actually save any space. We would really like to delete the account entirely (at least, on disk).

Reply:
I see your point. We don't really need to store the empty account though, we just need to act as if it were stored in the leaf. I will rephrase that bit.